File: static_files/assignments/HW2_files/SRC/MATLAB/svm_classifier_psth.m
function [accuracy_w, recall_manmade, recall_body, featuresMat , labels] = svm_classifier_psth(manmade_data, body_data, natural_data, face_data, number_of_time_slices, region)
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    % SVM Classifier on PSTH in 4 Categories
    % Inputs:
    %   - manmade_data: manmade PSTH data (samples x features x time slices)
    %   - body_data: Body PSTH data (samples x features x time slices)
    %   - natural_data: Natural PSTH data (samples x features x time slices)
    %   - face_data: Face PSTH data (samples x features x time slices)
    %   - number_of_time_slices: Number of time slices
    %   - region: Region label for saving results
    % Outputs:
    %   - accuracy_w: Accuracy for each time slice
    %   - recall_manmade: Recall for manmade category across time slices
    %   - recall_body: Recall for body category across time slices
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    iteration = 10; % Number of iterations
    out = struct(); % Initialize output struct
    accuracy_w = zeros(number_of_time_slices, 1); % Initialize accuracy vector

    % Perform SVM decoding for each time slice
    featuresMat =[];
    for ts = 1:number_of_time_slices
        % Prepare features for each category
        manmade_features = transpose(manmade_data(1:end, :, ts));
        body_features = transpose(body_data(1:end, :, ts));
        natural_features = transpose(natural_data(1:end, :, ts));
        face_features = transpose(face_data(1:end, :, ts));

        % Create labels
        manmade_labels = -1 * ones(size(manmade_features, 1), 1);
        body_labels = 1 * ones(size(body_features, 1), 1);
        natural_labels = -1 * ones(size(natural_features, 1), 1);
        face_labels = 1 * ones(size(face_features, 1), 1);

        % Stack features and labels
        features = [manmade_features; body_features; natural_features; face_features];
        labels = [manmade_labels; body_labels; natural_labels; face_labels];
        featuresMat = cat(3,featuresMat,features);
        % Perform SVM decoding
        disp(['Accuracy for time slice ' num2str(ts)]);
        out(ts).out = gen_fx_get_svm(labels, features, 0.5, iteration);
        accuracy_w(ts) = mean(out(ts).out.pt, 1) * 100;
    end

    % Time vector for plotting
    time = linspace(-200, 700, number_of_time_slices);

    % Plot accuracy
    figure();
    plot(time, accuracy_w, '-o', 'LineWidth', 5);
    hold on;
    plot(time, movmean(accuracy_w, 10), 'LineWidth', 5);
    grid on;
    legend({'Smooth Accuracy', 'Accuracy'}, 'FontSize', 16, 'Location', 'best');
    legend('boxoff');
    [t, s] = title(['Accuracy on rate ', region], 'Color', 'black');
    t.FontSize = 16;
    s.FontAngle = 'italic';
    xlabel('time (ms)', 'FontSize', 16, 'Color', 'b');
    ylabel('Percentage of Accuracy', 'FontSize', 16, 'Color', 'r');
    save(['Accuracy_animate_inanimate_rate_', region, '.mat'], 'accuracy_w', '-v7.3');

    % Recall calculations
    recall_manmade = zeros(1, number_of_time_slices);
    recall_body = zeros(1, number_of_time_slices);

    for ts = 1:number_of_time_slices
        confusion_mat = out(ts).out.C;
        recall_manmade_temp = zeros(1, iteration);
        recall_body_temp = zeros(1, iteration);

        for i = 1:iteration
            temp_confusion_matrix = confusion_mat(:, :, i);
            recall_manmade_temp(i) = temp_confusion_matrix(1, 1) / (temp_confusion_matrix(1, 1) + temp_confusion_matrix(1, 2));
            recall_body_temp(i) = temp_confusion_matrix(2, 2) / (temp_confusion_matrix(2, 1) + temp_confusion_matrix(2, 2));
        end

        recall_manmade(ts) = mean(recall_manmade_temp, 2) * 100;
        recall_body(ts) = mean(recall_body_temp, 2) * 100;
    end

    % Save recall data
    save(['Recall_animate_rate_', region, '.mat'], 'recall_manmade', '-v7.3');
    save(['Recall_inanimate_rate_', region, '.mat'], 'recall_body', '-v7.3');
    
    % Plot recall
    figure();
    plot(time, movmean(recall_manmade, 10), 'LineWidth', 5);
    hold on;
    plot(time, movmean(recall_body, 10), 'LineWidth', 5);
    grid on;
    legend({'Recall Inanimate', 'Recall Animate'}, 'FontSize', 16, 'Location', 'best');
    legend('boxoff');
    [t, s] = title(['Recall rate in ', region], 'Color', 'black');
    t.FontSize = 16;
    s.FontAngle = 'italic';
    xlabel('time (ms)', 'FontSize', 16, 'Color', 'b');
    ylabel('Percentage of Recall', 'FontSize', 16, 'Color', 'r');
end

File: static_files/assignments/HW2_files/SRC/Python/svm_classifier_psth.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import savemat
from sklearn.metrics import confusion_matrix
from gen_fx_get_svm import gen_fx_get_svm  # Make sure this is implemented correctly

def svm_classifier_psth(manmade_data, body_data, natural_data, face_data, number_of_time_slices, region):
    iteration = 10
    accuracy_w = np.zeros(number_of_time_slices)
    featuresMat = []
    out = []

    for ts in range(number_of_time_slices):
        manmade_features = manmade_data[:, :, ts].T
        body_features = body_data[:, :, ts].T
        natural_features = natural_data[:, :, ts].T
        face_features = face_data[:, :, ts].T

        manmade_labels = -1 * np.ones((manmade_features.shape[0], 1))
        body_labels = 1 * np.ones((body_features.shape[0], 1))
        natural_labels = -1 * np.ones((natural_features.shape[0], 1))
        face_labels = 1 * np.ones((face_features.shape[0], 1))

        features = np.vstack([manmade_features, body_features, natural_features, face_features])
        labels = np.vstack([manmade_labels, body_labels, natural_labels, face_labels]).ravel()
        featuresMat.append(features)

        print(f"Accuracy for time slice {ts+1}")
        result = gen_fx_get_svm(labels, features, 0.5, iteration)
        out.append(result)
        accuracy_w[ts] = np.mean(result['pt']) * 100

    time = np.linspace(-200, 700, number_of_time_slices)

    plt.figure()
    plt.plot(time, accuracy_w, '-o', linewidth=2)
    plt.plot(time, np.convolve(accuracy_w, np.ones(10)/10, mode='same'), linewidth=2)
    plt.grid(True)
    plt.title(f"Accuracy on rate {region}", fontsize=16, color='black')
    plt.xlabel("time (ms)", fontsize=16, color='blue')
    plt.ylabel("Percentage of Accuracy", fontsize=16, color='red')
    plt.legend(['Accuracy', 'Smooth Accuracy'], fontsize=14)
    plt.savefig(f"Accuracy_animate_inanimate_rate_{region}.png")
    savemat(f"Accuracy_animate_inanimate_rate_{region}.mat", {"accuracy_w": accuracy_w})

    recall_manmade = np.zeros(number_of_time_slices)
    recall_body = np.zeros(number_of_time_slices)

    for ts in range(number_of_time_slices):
        confusion_matrices = out[ts]['C']
        recall_manmade_temp = []
        recall_body_temp = []

        for i in range(iteration):
            cm = confusion_matrices[:, :, i]
            recall_manmade_temp.append(cm[0, 0] / (cm[0, 0] + cm[0, 1]))
            recall_body_temp.append(cm[1, 1] / (cm[1, 0] + cm[1, 1]))

        recall_manmade[ts] = np.mean(recall_manmade_temp) * 100
        recall_body[ts] = np.mean(recall_body_temp) * 100

    savemat(f"Recall_animate_rate_{region}.mat", {"recall_manmade": recall_manmade})
    savemat(f"Recall_inanimate_rate_{region}.mat", {"recall_body": recall_body})

    plt.figure()
    plt.plot(time, np.convolve(recall_manmade, np.ones(10)/10, mode='same'), linewidth=2)
    plt.plot(time, np.convolve(recall_body, np.ones(10)/10, mode='same'), linewidth=2)
    plt.grid(True)
    plt.title(f"Recall rate in {region}", fontsize=16, color='black')
    plt.xlabel("time (ms)", fontsize=16, color='blue')
    plt.ylabel("Percentage of Recall", fontsize=16, color='red')
    plt.legend(['Recall Inanimate', 'Recall Animate'], fontsize=14)
    plt.savefig(f"Recall_rate_{region}.png")

    return accuracy_w, recall_manmade, recall_body, np.stack(featuresMat, axis=2), labels
